File: app/src/main/java/fr/free/nrw/commons/MediaDataExtractor.java
package fr.free.nrw.commons;

import static fr.free.nrw.commons.depictions.Media.DepictedImagesFragment.PAGE_ID_PREFIX;

import androidx.core.text.HtmlCompat;
import fr.free.nrw.commons.media.Depictions;
import fr.free.nrw.commons.media.MediaClient;
import io.reactivex.Single;
import javax.inject.Inject;
import javax.inject.Singleton;
import org.jetbrains.annotations.NotNull;
import timber.log.Timber;

/**
 * Fetch additional media data from the network that we don't store locally.
 *
 * This includes things like category lists and multilingual descriptions,
 * which are not intrinsic to the media and may change due to editing.
 */
@Singleton
public class MediaDataExtractor {

  private final MediaClient mediaClient;

    @Inject
    public MediaDataExtractor(final MediaClient mediaClient) {
        this.mediaClient = mediaClient;
    }

    /**
     * Simplified method to extract all details required to show media details.
     * It fetches media object, deletion status, talk page and captions for the filename
     * @param filename for which the details are to be fetched
     * @return full Media object with all details including deletion status and talk page
     */
    public Single<Media> fetchMediaDetails(final String filename, final String pageId) {
      return Single.zip(getMediaFromFileName(filename),
            mediaClient.checkPageExistsUsingTitle("Commons:Deletion_requests/" + filename),
            getDiscussion(filename),
            pageId != null ? getCaption(PAGE_ID_PREFIX + pageId) : Single.just(""),
            getDepictions(filename),
            this::combineToMedia);
    }

  @NotNull
  private Media combineToMedia(final Media media, final Boolean deletionStatus, final String discussion,
      final String caption, final Depictions depictions) {
    media.setDiscussion(discussion);
    media.setCaption(caption);
    media.setDepictions(depictions);
    if (deletionStatus) {
        media.setRequestedDeletion(true);
    }
    return media;
  }

  /**
     * Obtains captions using filename
     * @param wikibaseIdentifier
     *
     * @return caption for the image in user's locale
     * Ex: "a nice painting" (english locale) and "No Caption" in case the caption is not available for the image
     */
    private Single<String> getCaption(final String wikibaseIdentifier) {
        return mediaClient.getCaptionByWikibaseIdentifier(wikibaseIdentifier);
    }

    /**
     * Fetch depictions from the MediaWiki API
     * @param filename the filename we will return the caption for
     * @return Depictions
     */
 private Single<Depictions> getDepictions(final String filename)  {
         return mediaClient.getDepictions(filename)
             .doOnError(throwable -> Timber.e(throwable, "error while fetching depictions"));
    }

    /**
     * Method can be used to fetch media for a given filename
     * @param filename Eg. File:Test.jpg
     * @return return data rich Media object
     */
    public Single<Media> getMediaFromFileName(final String filename) {
        return mediaClient.getMedia(filename);
    }

    /**
     * Fetch talk page from the MediaWiki API
     * @param filename
     * @return
     */
    private Single<String> getDiscussion(final String filename) {
        return mediaClient.getPageHtml(filename.replace("File", "File talk"))
                .map(discussion -> HtmlCompat.fromHtml(discussion, HtmlCompat.FROM_HTML_MODE_LEGACY).toString())
                .onErrorReturn(throwable -> {
                    Timber.e(throwable, "Error occurred while fetching discussion");
                    return "";
                });
    }
}

File: app/src/main/java/fr/free/nrw/commons/MediaDataExtractor.kt
package fr.free.nrw.commons

import androidx.core.text.HtmlCompat
import fr.free.nrw.commons.media.PAGE_ID_PREFIX
import fr.free.nrw.commons.media.IdAndCaptions
import fr.free.nrw.commons.media.MediaClient
import io.reactivex.Single
import timber.log.Timber
import javax.inject.Inject
import javax.inject.Singleton

/**
 * Fetch additional media data from the network that we don't store locally.
 *
 *
 * This includes things like category lists and multilingual descriptions, which are not intrinsic
 * to the media and may change due to editing.
 */
@Singleton
class MediaDataExtractor @Inject constructor(private val mediaClient: MediaClient) {

    fun fetchDepictionIdsAndLabels(media: Media) =
        mediaClient.getEntities(media.depictionIds)
            .map {
                it.entities()
                    .mapValues { entry -> entry.value.labels().mapValues { it.value.value() } }
            }
            .map { it.map { (key, value) -> IdAndCaptions(key, value) } }
            .onErrorReturn { emptyList() }

    fun checkDeletionRequestExists(media: Media) =
        mediaClient.checkPageExistsUsingTitle("Commons:Deletion_requests/" + media.filename)

    fun fetchDiscussion(media: Media) =
        mediaClient.getPageHtml(media.filename!!.replace("File", "File talk"))
            .map { HtmlCompat.fromHtml(it, HtmlCompat.FROM_HTML_MODE_LEGACY).toString() }
            .onErrorReturn {
                Timber.d("Error occurred while fetching discussion")
                ""
            }

    fun refresh(media: Media): Single<Media> {
        return Single.ambArray(
            mediaClient.getMediaById(PAGE_ID_PREFIX + media.pageId)
                .onErrorResumeNext { Single.never() },
            mediaClient.getMediaSuppressingErrors(media.filename)
                .onErrorResumeNext { Single.never() }
        )

    }

    fun getHtmlOfPage(title: String) = mediaClient.getPageHtml(title);

    /**
     * Fetches wikitext from mediaClient
     */
    fun getCurrentWikiText(title: String) = mediaClient.getCurrentWikiText(title);
}
